node-wasm/src/utils.rs

```diff
@@ -103,3 +106,79 @@ where
         })
     }
 }
+
+pub(crate) trait WorkerSelf {
+    type GlobalScope;
+
+    fn worker_self() -> Self::GlobalScope;
+}
+
+impl WorkerSelf for SharedWorker {
+    type GlobalScope = SharedWorkerGlobalScope;
+
+    fn worker_self() -> Self::GlobalScope {
+        JsValue::from(js_sys::global()).into()
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub(crate) enum JsResult<T, E>
+where
+    T: JsCast,
+    //E: Serialize + DeserializeOwned,
+{
+    #[serde(with = "serde_wasm_bindgen::preserve")]
+    Ok(T),
+    Err(E),
+}
+
+// once try_trait_v2 is stabilised, this can go
+impl<T, E> JsResult<T, E>
+where
+    T: JsCast,
+    E: Serialize + DeserializeOwned,
+{
+    pub fn into_result(self) -> Result<T, E> {
+        self.into()
+    }
+}
+
+impl<T, E> From<Result<T, E>> for JsResult<T, E>
+where
+    T: JsCast,
+    E: Serialize + DeserializeOwned,
+{
+    fn from(result: Result<T, E>) -> Self {
+        match result {
+            Ok(v) => JsResult::Ok(v),
+            Err(e) => JsResult::Err(e),
+        }
+    }
+}
+
+impl<T, E> From<JsResult<T, E>> for Result<T, E>
+where
+    T: JsCast,
+    E: Serialize + DeserializeOwned,
+{
+    fn from(result: JsResult<T, E>) -> Self {
+        match result {
+            JsResult::Ok(v) => Ok(v),
+            JsResult::Err(e) => Err(e),
+        }
+    }
+}
+
+impl<T, E> Debug for JsResult<T, E>
+where
+    T: JsCast,
+    E: Serialize + DeserializeOwned,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            JsResult::Ok(_) => f.debug_tuple("JsResult::Ok"),
+            JsResult::Err(_) => f.debug_tuple("JsResult::Err"),
```

Comment:
is there any reason not to require just `T & E: Debug` and include actual filelds?

Reply:
I wanted to be conservative with the `Debug` requirement on contained types, but ultimately it looks like it's not gonna be a problem.
Changed to full debug print 👍